src/devices/CharacterLcd/LcdInterface.I2c4Bit.cs

```diff
@@ -0,0 +1,137 @@
+            public override bool BacklightOn 
+            {
+                get
+                {
+                    return _backlightOn;
+                }
+                set
+                {
+                    _backlightOn = value;
+                    // Need to send a command to make this happen immediately.
+                    SendCommand(0);
+                }
+            }
+
+            private byte BacklightFlag
+            {
+                get
+                {
+                    if (BacklightOn)
+                    {
+                        return LCD_BACKLIGHT;
+                    }
+                    else
+                    {
+                        return 0;
+                    }
+                }
+            }
+
+            private void InitDisplay()
+            {
+                // This sequence (copied from a python example) completely resets the display (if it was 
+                // previously erroneously used with 8 bit access, it may not return to normal operation otherwise)
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x3);
+                SendCommandAndWait(0x2);
+                SendCommandAndWait(LCD_FUNCTIONSET | LCD_2LINE | LCD_5x8DOTS | LCD_4BITMODE);
+                SendCommandAndWait(LCD_DISPLAYCONTROL | LCD_DISPLAYON);
+                SendCommandAndWait(LCD_CLEARDISPLAY);
+                SendCommandAndWait(LCD_ENTRYMODESET | LCD_ENTRYLEFT);
+            }
+
+            private void SendCommandAndWait(byte command)
+            {
+                // Must not run the init sequence to fast or undefined behavior may occur
+                SendCommand(command);
+                Thread.Sleep(1);
+            }
+
+            public override void SendCommand(byte command)
+            {
+                Write4Bits((byte)(0x00 | (command & 0xF0)));
+                Write4Bits((byte)(0x00 | (command << 4 & 0xF0)));
+            }
+
+            private void Write4Bits(byte command)
+            {
+                _device.WriteByte((byte)(command | ENABLE | BacklightFlag));
+                _device.WriteByte((byte)((command & ~ENABLE) | BacklightFlag));
+            }
+
+            public override void SendCommands(ReadOnlySpan<byte> commands)
+            {
+                foreach (var c in commands)
+                {
+                    SendCommand(c);
+                }
+            }
+
+            public override void SendData(byte value)
+            {
+                Write4Bits((byte)(REGISTERSELECT | (value & 0xF0)));
+                Write4Bits((byte)(REGISTERSELECT | (value << 4 & 0xF0)));
+            }
+
+            public override void SendData(ReadOnlySpan<byte> values)
+            {
+                foreach (var c in values)
+                {
+                    Write4Bits((byte)(REGISTERSELECT | (c & 0xF0)));
+                    Write4Bits((byte)(REGISTERSELECT | (c << 4 & 0xF0)));
```

Comment:
Please add brackets here: `c << 4 & 0xF0` and above - otherwise it's unclear what order this will execute (I think `&` has higher priority)

Reply:
You are right, I'll clarify that. However << comes before & (see i.e. https://www.programiz.com/csharp-programming/operator-precedence-associativity); otherwise this code wouldn't work...